src/redisgears_python.c

```diff
@@ -2093,6 +2094,291 @@ static PyObject* createTensorFromValues(PyObject *cls, PyObject *args){
+        return NULL;
+    }
+    PyObject* keyName = PyTuple_GetItem(args, 0);
+    if(!PyUnicode_Check(keyName)){
+        PyErr_SetString(GearsError, "key name argument must be a string");
+        return NULL;
+    }
+
+    
+    PyTensor* pyt = (PyTensor*)PyTuple_GetItem(args, 1);
+    if(!PyObject_IsInstance((PyObject*)pyt, (PyObject*)&PyTensorType)){
+        PyErr_SetString(GearsError, "Given argument is not of type PyTensor");
+        return NULL;
+    }
+
+    RedisModuleType *ai_tensorType = RedisAI_TensorRedisType();
+    size_t len;
+    const char* keyNameCStr = PyUnicode_AsUTF8AndSize(keyName, &len);
+
+    if(Cluster_IsClusterMode()){
+        const char* keyNodeId = Cluster_GetNodeIdByKey(keyNameCStr);
+        if(!Cluster_IsMyId(keyNodeId)){
+            PyErr_SetString(GearsError, "Given key is not in the current shard");
+            return NULL;
+        }
+    }
+
+    PyObject* obj = NULL;
+    RedisModuleCtx* rctx = RedisModule_GetThreadSafeContext(NULL);
+    RedisModuleString *keyNameStr = RedisModule_CreateString(rctx, keyNameCStr, len);
+    LockHandler_Acquire(rctx);
+    RedisModuleKey *key = RedisModule_OpenKey(rctx,keyNameStr, REDISMODULE_WRITE);
+    // Check if the key is empty or 
+    bool empty = RedisModule_KeyType(key) == REDISMODULE_KEYTYPE_EMPTY;
+    if(!empty && (RedisModule_ModuleTypeGetType(key) != ai_tensorType)) {
+        PyErr_SetString(GearsError, "ERR Key is already set with non tensor values.");
+        goto clean_up;
+    }
+
+    // Set a shllow copy of the tensor.
+    RedisModule_ModuleTypeSetValue(key, ai_tensorType, RedisAI_TensorGetShallowCopy(pyt->t));
+    Py_INCREF(Py_None);
+    obj = Py_None;
+    
+clean_up:
+    RedisModule_CloseKey(key);
+    LockHandler_Release(rctx);
+    RedisModule_FreeString(rctx, keyNameStr);
+    RedisModule_FreeThreadSafeContext(rctx);
+    return obj;
+}
+
+static PyObject* msetTensorsInKeyspace(PyObject *cls, PyObject *args) {
+    verifyRedisAILoaded();
+    if(PyTuple_Size(args) != 1){
+        PyErr_SetString(GearsError, "Wrong number of arguments given to msetTensorInKeyspace");
+        return NULL;
+    }
+    // Input should be a Dictionary<string, tensor>.
+    PyObject* py_tensorDict = (PyListObject*)PyTuple_GetItem(args, 0);
+    if(!PyDict_Check(py_tensorDict)){
+        PyErr_SetString(GearsError, "Given argument is not of type PyDict");
+        return NULL;
+    }
+
+    PyObject* obj = NULL;
+    bool error = false;
+    bool is_cluster = Cluster_IsClusterMode();
+    size_t len = PyDict_Size(py_tensorDict);
+    size_t stack_buffer_size = 1000;
+    array_new_on_stack(RedisModuleKey*, stack_buffer_size, keys);
+    array_new_on_stack(RAI_Tensor*, stack_buffer_size, tensorList);
+    array_new_on_stack(const char*, stack_buffer_size, keyNameList);
+
+    Py_ssize_t i = 0;
+    PyObject *key, *value;
+    // Validate each key-value pair for <string, tensor> type.
+    while (PyDict_Next(py_tensorDict, &i, &key, &value)) {
```

Comment:
What the difference between this and get iterator ? Could we still run on this dictionary after?

Reply:
I saw this example at `dictobject.c`. I believe this is a managed iteration over the data structure so. As far as I understand the code it depends on the value of `i` so It is reusable.

Reply:
Ohh right I missed the i, :+1: